samples/sample_common/src/vaapi_utils_drm.cpp

```diff
@@ -22,169 +22,93 @@ or https://software.intel.com/en-us/media-client-solutions-support.
-                        adapters[adapters_num].device_id = device_id;
-
-                        ++adapters_num;
-                    }
-                }
-            }
+            adapters.push_back(i);
         }
-        free(dir_entries[i]);
+        close(fd);
     }
-    if (entries_num) free(dir_entries);
-    if (p_adapters) *p_adapters = adapters;
-
-    return adapters_num;
 }
 
 DRMLibVA::DRMLibVA(int type)
     : CLibVA(type)
     , m_fd(-1)
 {
-    const mfxU32 IntelVendorID = 0x8086;
-    //the first Intel adapter is only required now, the second - in the future
+    // The first Intel adapter is only required now, the second - in the future
     const mfxU32 numberOfRequiredIntelAdapter = 1;
-    const char nodesNames[][8] = {"renderD", "card"};
 
     VAStatus va_res = VA_STATUS_SUCCESS;
     mfxStatus sts = MFX_ERR_NONE;
     int major_version = 0, minor_version = 0;
 
-    mfx_disp_adapters* adapters = NULL;
-    int adapters_num = mfx_init_adapters(&adapters);
-
-    // Search for the required display adapter
-    int i = 0, nFoundAdapters = 0;
-    int nodesNumbers[] = {0,0};
-    while ((i < adapters_num) && (nFoundAdapters != numberOfRequiredIntelAdapter))
-    {
-        if (adapters[i].vendor_id == IntelVendorID)
-        {
-            nFoundAdapters++;
-            nodesNumbers[0] = i+128; //for render nodes
-            nodesNumbers[1] = i;     //for card
-        }
-        i++;
-    }
-    if (adapters_num) free(adapters);
-    // If Intel adapter with specified number wasn't found, throws exception
-    if (nFoundAdapters != numberOfRequiredIntelAdapter)
-        throw std::range_error("The Intel adapter with a specified number wasn't found");
-
-    // Initialization of paths to the device nodes
-    char** adapterPaths = new char* [2];
-    for (int i=0; i<2; i++)
-    {
-        if ((i == 0) && (type == MFX_LIBVA_DRM_MODESET)) {
-          adapterPaths[i] = NULL;
-          continue;
-        }
-        adapterPaths[i] = new char[sizeof(MFX_DRI_DIR) + sizeof(nodesNames[i]) + 3];
-        sprintf(adapterPaths[i], "%s%s%d", MFX_DRI_DIR, nodesNames[i], nodesNumbers[i]);
+    std::vector<mfxU32> adapters;
+    get_intel_adapters(adapters);
+    
+    if (adapters.size() != numberOfRequiredIntelAdapter)
+        throw std::range_error("Intel GPU was not found");
+
+    // Initialize device paths using first found Intel adapter
+    std::string adapterPaths[MFX_DRI_NODE_NUM];
+    if (type != MFX_LIBVA_DRM_MODESET) {
+        adapterPaths[MFX_DRI_NODE_RENDER] = MFX_DRI_NODE_RENDER_PATH;
+        adapterPaths[MFX_DRI_NODE_RENDER] += std::to_string(MFX_DRI_RENDER_START_INDEX + adapters.front());
     }
+    adapterPaths[MFX_DRI_NODE_CARD] = MFX_DRI_NODE_CARD_PATH;
+    adapterPaths[MFX_DRI_NODE_CARD] += std::to_string(adapters.front());
 
     // Loading display. At first trying to open render nodes, then card.
-    for (int i=0; i<2; i++)
+    for (int i = MFX_DRI_NODE_RENDER; i < MFX_DRI_NODE_NUM; i++)
```

Comment:
this cycle is again example of tricky logic without a need in it. But that wasn't introduced by you, so I'm ok to leave it as is, but this is next candidate for refactoring